pkg/goldpinger/client.go

```diff
@@ -273,7 +274,12 @@ func getClient(hostIP string) (*apiclient.Goldpinger, error) {
 	if hostIP == "" {
 		return nil, errors.New("Host or pod IP empty, can't make a call")
 	}
-	host := fmt.Sprintf("%s:%d", hostIP, GoldpingerConfig.Port)
+	var host string
+	if k8snet.IsIPv6String(hostIP) {
+		host = fmt.Sprintf("[%s]:%d", hostIP, GoldpingerConfig.Port)
+	} else {
+		host = fmt.Sprintf("%s:%d", hostIP, GoldpingerConfig.Port)
+	}
```

Comment:
Hey @rbtr thanks for chiming in!

> Stringly typing something that is as well defined as an IP smells to me, so I went looking to see what Go provides for dealing with them in this context and found: [net.JoinHostPort](https://pkg.go.dev/net#JoinHostPort).
> 
> Instead of continuing to pass around IP strings, I think part of this change should be to ingest the IP string from the pod immediately in to a `net.IP` (see: [net.ParseIP](https://pkg.go.dev/net#JoinHostPort)). Then, when we use it, we don't care if it's v4 or v6 - the net package will do the right thing for us. As a side-effect, you might then be able to remove (at least significantly simplify) the changes made in k8s.go.

Agreed. That would definitely be a welcome improvement. That said, I'd argue that's probably best addressed by a separate PR as a refactor.


> If we then deal with IPs the same way regardless of v4/v6, I'm not entirely convinced that picking v4/v6 should be a runtime option. Since on a v4 cluster we only have v4 addresses, and a dual stack cluster pods always have v4 and v6 addresses, it might make more sense to always ping all the pod IPs - then goldpinger doesn't _really_ need to know about v4 vs v6.

I see your logic here, but I think that being able to pick v4 vs v6 vs v4+v6 might still have merit, depending on what you're after. For example, @tyler-lloyd suggestion to have two separate deployments, a v4 and v6 one, might actually be easier to understand/auto-validate. We'd also need to tweak the UI to allow for multiple IPs per host/pod too.

@tyler-lloyd @paulgmiller @rbtr how would you feel about turning `USE_IPV6` into `IP_VERSIONS` that's a list of versions to ping, that defaults to `4` for backwards compatibility? This would allow for all scenarios, methinks.

Reply:
> how would you feel about turning USE_IPV6 into IP_VERSIONS that's a list of versions to ping

sure, I like that better. Are you suggesting put that in this PR or future work? We could change the argument to IP_VERSIONS and only allow IPv4 or IPv6 for now and come back to add IPv4,IPv6 functionality and don't need to add or update the arguments.

Reply:
That sounds good to me. 👍

Reply:
@seeker89  updated with `IP_VERSIONS` and will default to "4" and for now only use the first entry (i.e. if you put "4", "6" it is only going to use "4")